### PR TITLE
Camera capture feature

### PR DESCRIPTION
**Pull request by JuanLeee**
Working version of camera capturing feature with resolution and position changeability.
Needs work with orientation when rotating device.

### CODE REVIEW

File: Sources/SwiftProcessing/Core/UIKitElements/Capture.swift [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+        }
+        return controller
+    }
+    
+    class var topViewController: UIViewController? { return topViewController() }
+}
+
+extension Sketch{
+    
+    open func createCamera(position: String) -> Camera{
```
Comment (jjkaufman) [inline Sources/SwiftProcessing/Core/UIKitElements/Capture.swift:293]: add the possible positions as constants to the main Sketch class. That way createCamera could be called like so

createCamera(FRONT)
createCamera(BACK)

File: Sources/SwiftProcessing/Core/UIKitElements/Capture.swift [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+        let b = Camera(self)
+        b.prepare(cameraPosition: b.getCameraPosition(position: position)) { success in
+            if success { b.start() }
```
Comment (jjkaufman) [inline Sources/SwiftProcessing/Core/UIKitElements/Capture.swift:296]: what are the possible failure states here? At a minimum, add some console prints to help the user know something went wrong during the preparation

File: Sources/SwiftProcessing/Core/UIKitElements/Capture.swift [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+            self.captureCompletionBlock = nil
+        }
+    }
+}
+
+// Navigation
+
+extension Camera {
+    
+    private func show(alert: UIAlertController) {
```
Comment (jjkaufman) [inline Sources/SwiftProcessing/Core/UIKitElements/Capture.swift:215]: the show functions + topViewController extension would make sense in another file named Alerts. I can see them being useful for other elements in the future.

In Camera, you could instead create an instance of your Alerts object and leverage from the functions from there.

File: Sources/SwiftProcessing/Core/UIKitElements/Capture.swift [hunk 4]
```diff
@@ -0,0 +1,310 @@
+import UIKit
+import AVFoundation
+import Vision
+
+open class Camera: UIKitViewElement {
```
Comment (jjkaufman) [inline Sources/SwiftProcessing/Core/UIKitElements/Capture.swift:5]: run the auto format on any file edited in this PR. Some tabbing seems off. 

In xCode, select all the code in the file, and hit control + i

File: Sources/SwiftProcessing/Core/UIKitElements/Capture.swift [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+                            "VGA" : AVCaptureSession.Preset.vga640x480,
+                            "720" : AVCaptureSession.Preset.hd1280x720,
+                            "1080" : AVCaptureSession.Preset.hd1920x1080]
+    
+    init(_ view: Sketch) {
+        self.previewView = UIView()
+        super.init(view, self.previewView!)
+    }
+    
+    open func setResolution(resolution: String) {
```
Comment (jjkaufman) [inline Sources/SwiftProcessing/Core/UIKitElements/Capture.swift:41]: add the possible resolutions as constants to the main Sketch class. That way setResolution could be called like so

setResolution(HIGH)

File: Sources/SwiftProcessing/Core/UIKitElements/Capture.swift [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+            
+        } catch {
+            show(error: error as NSError)
+            return
+        }
+        
+        // Create the video data output
+        let videoOutput = AVCaptureVideoDataOutput()
+        videoOutput.setSampleBufferDelegate(self, queue: dataOutputQueue)
+        videoOutput.videoSettings = [kCVPixelBufferPixelFormatTypeKey as String: kCVPixelFormatType_32BGRA]
```
Comment (jjkaufman) [inline Sources/SwiftProcessing/Core/UIKitElements/Capture.swift:180]: curious what types of video settings are available here? May be more items to research and think about for configuration

File: Sources/SwiftProcessing/Core/UIKitElements/Capture.swift [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+        
+        // Add the video output to the capture session
+        session.addOutput(videoOutput)
+        
+        let videoConnection = videoOutput.connection(with: .video)
+        videoConnection?.videoOrientation = .portrait
```
Comment (jjkaufman) [inline Sources/SwiftProcessing/Core/UIKitElements/Capture.swift:186]: I'm thinking it makes sense to make this the devices current orientation. You can read that using 

UIDevice.current.orientation.isLandscape

File: Sources/SwiftProcessing/Core/UIKitElements/Capture.swift [hunk 8]
```diff
@@ -0,0 +1,310 @@
+import UIKit
+import AVFoundation
+import Vision
```
Comment (jjkaufman) [inline Sources/SwiftProcessing/Core/UIKitElements/Capture.swift:3]: is this import needed?

Comment (JuanLeee) [reply Sources/SwiftProcessing/Core/UIKitElements/Capture.swift:3]: Yes is required for line 12, sequence handler

File: Sources/SwiftProcessing/Core/UIKitElements/Capture.swift [hunk 9]
```diff
@@ -0,0 +1,310 @@
...
+    func setPhoto(finished: @escaping () -> Void) {
+        self.capturePhoto { image in
+            let newPhoto = self.flipImageLeftRight(image)
+            self.photo =  Image(newPhoto!)
+            finished()
+        }
+        
+    }
+    
+    open func get() -> Image? {
```
Comment (jjkaufman) [inline Sources/SwiftProcessing/Core/UIKitElements/Capture.swift:76]: param inputs still needed

File: Sources/SwiftProcessing/Core/UIKitElements/Capture.swift [hunk 10]
```diff
@@ -0,0 +1,385 @@
...
+        guard let range = self.cameraDevice!.formats.first!.videoSupportedFrameRateRanges.first,
+            range.minFrameRate...range.maxFrameRate ~= (desiredFrameRate)
+            else {
+                print("Requested FPS is not supported by the device's activeFormat !")
+                return
+        }
+        
+        do {
+            try self.cameraDevice!.lockForConfiguration()
+            self.cameraDevice!.activeVideoMinFrameDuration = CMTimeMake(value: 1,timescale: Int32(desiredFrameRate))
```
Comment (jjkaufman) [inline Sources/SwiftProcessing/Core/UIKitElements/Capture.swift:74]: if min is set to a lower value, do we gain any performance?

File: Sources/SwiftProcessing/Core/UIKitElements/Capture.swift [hunk 11]
```diff
@@ -0,0 +1,385 @@
...
+            return AVCaptureDevice.Position.front
+        }
+    }
+    
+    func setPhoto(_ width: CGFloat? = nil,_ height: CGFloat? = nil,_ finished: @escaping () -> Void) {
+        self.capturePhoto { image in
+            // let newPhoto = self.flipImageLeftRight(image)
+            self.photo =  Image(image)
+            if width != nil && height != nil {
+                self.photo!.resize(width!,height!)
```
Comment (jjkaufman) [inline Sources/SwiftProcessing/Core/UIKitElements/Capture.swift:105]: photo.get instead of resize

Comment (jjkaufman) [reply Sources/SwiftProcessing/Core/UIKitElements/Capture.swift:105]: get takes in an x, y, w, h

File: Sources/SwiftProcessing/Core/UIKitElements/Capture.swift [hunk 12]
```diff
@@ -0,0 +1,385 @@
...
+            }
+            finished()
+        }
+        
+    }
+    
+    open func get(_ width: CGFloat? = nil,_ height: CGFloat? = nil) -> Image? {
```
Comment (jjkaufman) [inline Sources/SwiftProcessing/Core/UIKitElements/Capture.swift:112]: add x, y

File: Sources/SwiftProcessing/Core/UIKitElements/Capture.swift [hunk 13]
```diff
@@ -0,0 +1,385 @@
...
+        setPhoto(width,height) {}
+        return self.photo
+    }
+    
+    func flipImageLeftRight(_ image: UIImage) -> UIImage? {
```
Comment (jjkaufman) [inline Sources/SwiftProcessing/Core/UIKitElements/Capture.swift:117]: not used anymore, lets delete this

File: Sources/SwiftProcessing/Core/UIKitElements/Capture.swift [hunk 14]
```diff
@@ -0,0 +1,385 @@
...
+            if camera.isFocusModeSupported(.continuousAutoFocus) {
+                camera.focusMode = .continuousAutoFocus
+            }
+            
+            if camera.isExposureModeSupported(.continuousAutoExposure) {
+                camera.exposureMode = .continuousAutoExposure
+            }
+            if desiredFrameRate != nil {
+                
+                guard let range = camera.formats.first!.videoSupportedFrameRateRanges.first,
```
Comment (jjkaufman) [inline Sources/SwiftProcessing/Core/UIKitElements/Capture.swift:210]: 210 -> 226 move that to private function named something like configureFrameRate

File: Sources/SwiftProcessing/Core/UIKitElements/Capture.swift [hunk 15]
```diff
@@ -0,0 +1,385 @@
...
+                                      space: colorSpace, bitmapInfo: bitmapInfo.rawValue) else { return nil }
+        
+        guard let cgImage = context.makeImage() else { return nil }
+        self.init(cgImage: cgImage, scale: 1, orientation: orientation)
+    }
+}
+
+///////////////////////////////////////////////////////////////////////////
+
+extension UIApplication {
```
Comment (jjkaufman) [inline Sources/SwiftProcessing/Core/UIKitElements/Capture.swift:347]: move to UIApplicationExtensions file

File: Sources/SwiftProcessing/Core/UIKitElements/Capture.swift [hunk 16]
```diff
@@ -0,0 +1,385 @@
...
+        let b = Camera(self)
+        b.prepare(cameraPosition: b.getCameraPosition(position), desiredFrameRate: desiredFrameRate) { success in
+            if success { b.start() }
+        }
+        viewRefs[b.id] = b
+        return b
+    }
+    
+    open func createCamera(_ desiredFrameRate: Int? = nil) -> Camera{
+        let b = Camera(self)
```
Comment (jjkaufman) [inline Sources/SwiftProcessing/Core/UIKitElements/Capture.swift:378]: call the function above

File: Sources/SwiftProcessing/Core/UIKitElements/Capture.swift [hunk 17]
```diff
@@ -0,0 +1,385 @@
...
+        if let AVPosition = self.AVCapturePositions[position] {
+            return AVPosition
+        } else {
+            return AVCaptureDevice.Position.front
+        }
+    }
+    
+    func setPhoto(_ width: CGFloat? = nil,_ height: CGFloat? = nil,_ finished: @escaping () -> Void) {
+        self.capturePhoto { image in
+            // let newPhoto = self.flipImageLeftRight(image)
```
Comment (jjkaufman) [inline Sources/SwiftProcessing/Core/UIKitElements/Capture.swift:102]: remove comment code

File: Sources/SwiftProcessing/Core/UIKitElements/Capture.swift [hunk 18]
```diff
@@ -0,0 +1,347 @@
...
+            self.photo =  Image(image)
+            if width != nil && height != nil {
+                self.photo = self.photo!.get(x,y,width!,height!)
+            }
+            finished()
+        }
+        
+    }
+    
+    open func get(_ width: CGFloat? = nil,_ height: CGFloat? = nil,_ x: CGFloat = 0,_ y: CGFloat = 0) -> Image? {
```
Comment (jjkaufman) [inline Sources/SwiftProcessing/Core/UIKitElements/Capture.swift:110]: last comment :) 

Lets switch this around to match the order of other functions. x, y, width, height

**Pull request merged**